### PR TITLE
Implement programs page

### LINKED ISSUE #259: Build Programs Page
https://www.figma.com/file/kKaWvJvyzhNS9hD21v7Ei5/ShelterTech.org-2020?node-id=4216%3A0

This will also include building the new component on that page with the individual program listings.

Due date: 7/15

### PR DESCRIPTION
**Pull request by ivanleeswe**
Fix #259

### CODE REVIEW

File: src/components/grid-aware/ArticleSpotlightCard/ArticleSpotlightCard.module.css
Comment (richardxia) [inline src/components/grid-aware/ArticleSpotlightCard/ArticleSpotlightCard.module.css]: I think this change shouldn't be here, right? The Article Spotlight card doesn't even appear on the new Programs page, and adding this extra margin makes the bottom of the Home Page look like it has too much space.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: Ah, all three of these blocks (SF Service Guide, ShelterConnect, and Community Rep) should just use the same, single component definition but with different props, since they all have an identical visual structure. Although the main description may have a different number of paragraphs or bulleted lists, we can actually just represent that as a single `description` prop that accepts a JSX fragment, which would allow us to pass in different sets of `<p>`, `<ul>`, and other elements.

You can take a look at how we do this for the ImageHeader component as instantiated in the new [About Us page](https://github.com/ShelterTechSF/sheltertech.org/blob/dfd799813a656f4fd3be125e704a75b943fcbe4b/src/pages/about/index.tsx#L23-L68) and on the [Volunteer page](https://github.com/ShelterTechSF/sheltertech.org/blob/dfd799813a656f4fd3be125e704a75b943fcbe4b/src/pages/volunteer/index.tsx#L40).

Do you think you could go consolidate these into a single component and just provide them different props on the programs page? We could call the component the `ProgramBlock` or something like that.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: Let's have all the donate buttons on the page take you to the Donate page.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: Can we have this pop up the same partnership modal as on the Home Page? It's the one in the "Get Involved" section, under the "Partnerships" subsection.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: Let's have this be an external link that takes you to https://sfserviceguide.org/.

**Review by richardxia — COMMENTED** [review]
Thanks for working on this! I think my main comment is that we should be defining a single React component that can be used with all three of the programs listed on the page, so I'm going to hold off from reviewing the three separate subcomponents that you have here for now. That said, I think you've done a great job faithfully implementing the designs, so I don't expect to have too many follow up comments about that once the components have been merged into one.

A few other comments I'd like to make are:

- Could you merge in the latest `main` branch? You currently have a conflict due to @candywang's About Us page being merged in and also touching the same part of the Navigation component
- Could you also update the footer navigation such that the Programs link is renamed to Our Programs and links to this new page?

File: src/components/grid-aware/ArticleSpotlightCard/ArticleSpotlightCard.module.css
Comment (ivanleeswe) [reply src/components/grid-aware/ArticleSpotlightCard/ArticleSpotlightCard.module.css]: Strange, don't know how that got there. Deleted.

File: src/components/layout.tsx
Comment (richardxia) [inline src/components/layout.tsx]: Could you update the link text to say "Our Programs"? You made the change for the footer, but not the top navigation.

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.tsx
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.tsx]: I think we could probably give this a better name, since this is more than just a paragraph, as it also has a heading. Since this plays a similar role as the `ImageHeader` and `VideoHeader` components — being the top-level heading of the page — what if we called this the `TextHeader`?

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.tsx
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.tsx]: Since this definitely plays the role of being the top-level heading of the page, let's use the `<h1>` tag for this element. Note that you may need to update the CSS to override some default styling that comes from the `<h1>`.

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.stories.jsx
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.stories.jsx]: Nit: This is much longer than the actual title of the actual Programs page, so much that I think it's a bit misleading because the rendered component in Storybook doesn't resemble the actual component on the Programs page. Could we shrink this down?

Here's what it looks like between the Storybook example and the actual Our Programs page, where the ratio between the area of the title and the area of the description is completely inverted:

## Storybook
Screen Shot 2021-08-02 at 9 15 16 PM

## Actual page
Screen Shot 2021-08-02 at 9 15 11 PM

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css]: Is there a reason this is here? It looks like you only override this in both the title and the description to set it back to the normal left-alignment.

If this doesn't have a purpose, can you delete this and the left-alignment overrides below?

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css]: Nit: Put a space between the `2` and the `/` so that there's consistent spacing around the `/`.

```suggestion
  grid-column: 2 / span 6;
```

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css]: You only have one grid row (both the title and the description are within the same grid cell), so you can remove this property that declares that there are two rows.

File: src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramsPageLargeParagraph/ProgramsPageLargeParagraph.module.css]: Nit: Could you add a blank line between CSS blocks?

```suggestion
}

/* Override grid item default of auto so that item contents cannot stretch the grid itself */
```

File: src/components/grid-aware/ProgramsSubTitleBlock/ProgramsSubTitleBlock.tsx
Comment (richardxia) [inline src/components/grid-aware/ProgramsSubTitleBlock/ProgramsSubTitleBlock.tsx]: Nit: "Subtitle" is a single English word, so let's leave the "t" in title lowercase in the name of this component (and all the file names).

File: src/components/grid-aware/ProgramsSubTitleBlock/ProgramsSubTitleBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramsSubTitleBlock/ProgramsSubTitleBlock.module.css]: Instead of making this padding, let's make it margin. Padding is the spacing that's "inside" of an element, while margin is the spacing that's "outside" of the element. When we don't have a border or background color, it's not as apparent visually, but it's still important to think about whether spacing is internal or external to a component.

File: src/components/grid-aware/ProgramsSubTitleBlock/ProgramsSubTitleBlock.tsx
Comment (richardxia) [inline src/components/grid-aware/ProgramsSubTitleBlock/ProgramsSubTitleBlock.tsx]: Since this is found lower on the page, this shouldn't be an `<h1>` tag, which should be reserved for the page title, of which there should only be one on the page. Could you change this to be an `<h2>`?

File: src/components/grid-aware/ProgramBlock/ProgramBlock.tsx
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.tsx]: It's a bit weird to name this type `ServiceGuideProps`, given that it has nothing to do with the Service Guide. Did you mean to name this `ImageProps`?

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: Instead of reversing the flex direction, I think it'd be better if you just put the image before the text in the HTML. We should avoid doing unusual things like using `column-reverse` since it changes the layout order to not match the HTML order, which makes things a bit more tricky to read.

In addition, you may not even need to use flexbox at all, since if you're using a column flex direction and not using other flexbox-specific features, then this isn't really any different than `display: block`.

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: Nit: Let's put a space between the `6` and the `/`.

```suggestion
  grid-column: 6 / span 6;
```

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css:69]: I see that for several of the elements here, in order to space them apart you apply bottom margin, but for others you apply top margin. There's tradeoffs between the two approaches, but I generally find that it's a good practice to pick one direction and apply it consistently; that is, either primarily use bottom margin to separate components or primarily use top margin to separate components.

Could you pick a direction and make it consistent?

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: This style doesn't seem to be necessary, since the non-media-query version of it also sets a top margin of 0. Could you delete this?

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: You should probably use the direct child selector, `>`, here so that you don't select all first children in the whole tree. For example, not only does this select the first child of `.description`, but if you have a `<ul><li>Item 1</li><li>Item 2</li></ul>`, then it'll also select that first `<li>` because it's the first child of `<ul>`.

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: Setting a margin of 0 might be too extreme here. I think you are just looking to eliminate the top margin of the first element. Since any arbitrary HTML element can appear here, you don't want to overly override styles for it.

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: Three comments:

1. The default list-style-position is outside, so you don't need to explicitly set it here
2. You can just use the [`list-style`](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style) shorthand property to set of the three specific `list-style-*` subproperties
3. You can define these styles on the `ul` element, which is the more common way of doing it unless if you really needed to set a different style on some, but not all `li` elements (which would be pretty unusual).

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: I _really_ want to say that we should just use the new `column-gap` property for flexbox here, since that'll allow us to avoid having to do this annoying `:not(:last-child)` selector. Originally, the [`gap`](https://developer.mozilla.org/en-US/docs/Web/CSS/gap) properties were only for CSS Grid, but it was recently updated so that they could apply to flexbox as well. Firefox started supporting gap for flexbox in 2018, and Chrome and Edge added it in 2020, but Safari only added support a few months ago. Overall, it looks like only 82% of users are on browser versions that support it: https://caniuse.com/flexbox-gap

How about we use `column-gap`, and we get bug reports from people on older browsers, then we go back to this `:not(:last-child)` approach?

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: This comment doesn't seem right. The image is on the left side.

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: These two properties by themselves are going to stretch and distort the images. You'll want to add [`object-fit: cover`](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) so that the images are cropped to fit instead of stretched.

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: I'm assuming that you got this from the Figma design, but I think that was not meant to be precise, so let's just round this up to 180px.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: For this and all the other CTA links to the Donate page, could you use an internalLink instead of an externalLink? Making it an external link causes it to open in a new tab instead of directly taking you there.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: I think you have an older version of this image, where the image is cropped in a way where the Volunteer Spotlight box obscures her face. @candywang's very recent PR, https://github.com/ShelterTechSF/sheltertech.org/pull/267, updated the image such that Laura's face is further to the right, so could you pull in her updates and make sure you use this updated image?

**Review by richardxia — COMMENTED** [review]
Thanks for making the changes to this PR! The consolidated ProgramBlock component is exactly what I was looking for, so thanks for making that change.

I've now gone through and reviewed the individual components in a lot more detail. I have a lot of little comments, but there's nothing major, so I think this is in pretty good shape to get merged in soon. Thanks again for working on this!

File: src/components/grid-aware/ProgramBlock/ProgramBlock.module.css
Comment (richardxia) [inline src/components/grid-aware/ProgramBlock/ProgramBlock.module.css]: Let's delete this commented out style if it's no longer being used at all.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: Can you leave off the `https://www.sheltertech.org` from all these internal links? Having the full domain here makes the link behave strangely in local development environments, since clicking on the link causes it to open the real, production site instead of navigating there in the local environment. This would also affect staging, as well as any version of the site that's not the production one.

There's two more `internalLink`s lower on the page that should be fixed up as well.

File: src/pages/programs/index.tsx
Comment (richardxia) [reply src/pages/programs/index.tsx]: I just wanted to bump this comment, since it still looks like you're using the older image. You can tell the difference because in the older one, the black box containing the "Volunteer Spotlight" text overlaps her face, while the newer one doesn't.

File: src/pages/programs/index.tsx
Comment (richardxia) [inline src/pages/programs/index.tsx]: What's up with this .37px spacing on mobile? If it shows up in the Figma design, then I think we can assume that it wasn't intentional, and that we can just round this to 0.

**Review by richardxia — COMMENTED** [review]
Nice, this is getting close to being ready to merge. I have a few more comments for the latest batch of changes, and I wanted to bring up an older comment I made about using the newer photo of Laura for the Volunteer Spotlight component, since it looks like we're still using the older one.

Also, I just saw this image that doesn't appear to be used: src/pages/programs/mission-hotel.jpeg. Could you remove it if it's not in use?

**Review by richardxia — APPROVED** [review]
LGTM!

**Pull request merged**
